beacon_node/http_api/src/publish_blocks.rs

```diff
@@ -31,30 +34,89 @@ pub async fn publish_block<T: BeaconChainTypes>(
     chain: Arc<BeaconChain<T>>,
     network_tx: &UnboundedSender<NetworkMessage<T::EthSpec>>,
     log: Logger,
+    validation_level: BroadcastValidation,
 ) -> Result<(), Rejection> {
     let seen_timestamp = timestamp_now();
     let (block, is_locally_built_block) = match provenanced_block {
         ProvenancedBlock::Local(block) => (block, true),
         ProvenancedBlock::Builder(block) => (block, false),
     };
     let delay = get_block_delay_ms(seen_timestamp, block.message(), &chain.slot_clock);
+    debug!(log, "Signed block received in HTTP API"; "slot" => block.slot());
 
-    debug!(
-        log,
-        "Signed block published to HTTP API";
-        "slot" => block.slot()
-    );
+    let block_root = block_root.unwrap_or_else(|| block.canonical_root());
 
-    // Send the block, regardless of whether or not it is valid. The API
-    // specification is very clear that this is the desired behaviour.
+    /* actually publish a block */
+    let publish_block = move |block: Arc<SignedBeaconBlock<T::EthSpec>>,
+                              sender,
+                              log,
+                              seen_timestamp| {
+        let publish_timestamp = timestamp_now();
+        let publish_delay = publish_timestamp
+            .checked_sub(seen_timestamp)
+            .unwrap_or_else(|| Duration::from_secs(0));
 
-    let message = PubsubMessage::BeaconBlock(block.clone());
-    crate::publish_pubsub_message(network_tx, message)?;
+        info!(log, "Signed block published to network via HTTP API"; "slot" => block.slot(), "publish_delay" => ?publish_delay);
 
-    let block_root = block_root.unwrap_or_else(|| block.canonical_root());
+        let message = PubsubMessage::BeaconBlock(block);
+        crate::publish_pubsub_message(&sender, message)
+            .map_err(|_| BlockError::<T::EthSpec>::PublishError)
+    };
+
+    /* if we can form a `GossipVerifiedBlock`, we've passed our basic gossip checks */
+    let gossip_verified_block = GossipVerifiedBlock::new(block.clone(), chain.clone().as_ref()).map_err(|e| {
```

Comment:
Here is where we'd use the new trait `IntoGossipVerifiedBlock`, by calling `block.into_gossip_verified_block(&chain).map_err(...)`